File: src/main/java/com/developerphil/adbidea/ui/BoundTableModel.java
package com.developerphil.adbidea.ui;

import com.developerphil.adbidea.bean.BoundDataType;
import com.developerphil.adbidea.bean.BoundItemBean;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import javax.swing.table.AbstractTableModel;
import org.apache.commons.lang3.StringUtils;

public class BoundTableModel extends AbstractTableModel{
    private Class[] cellType = { Boolean.class, String.class, String.class, BoundDataType.class };
    private String title[] = { "select", "key", "value" ,"type"};

    private List<BoundItemBean> data = new ArrayList<>();

    public BoundTableModel() {
    }


    public List<BoundItemBean> getData() {
        return data.stream().filter(BoundItemBean -> BoundItemBean.getSelected()&&StringUtils.isNotEmpty(BoundItemBean.getKey())).collect(Collectors.toList());
    }

    public void setData(List<BoundItemBean> data) {
        this.data.clear();
        this.data.addAll(data);
        fireTableDataChanged();
    }

    public void clear() {
        int size = this.data.size();
        this.data.clear();
        fireTableRowsDeleted(0, size);
    }

    public void addEmptyRow() {
        data.add(new BoundItemBean(true, "", "",BoundDataType.NULL));
        fireTableRowsInserted(data.size() - 1, data.size() - 1);
    }

    public void removeRow(int row) {
        if (row > -1 && row < data.size()) {
            data.remove(row);
        }
        fireTableRowsDeleted(row, row);
    }

    @Override
    public Class<?> getColumnClass(int arg0) {
        return cellType[arg0];
    }

    @Override
    public String getColumnName(int arg0) {
        return title[arg0];
    }

    @Override
    public int getColumnCount() {
        return title.length;
    }

    @Override
    public int getRowCount() {
        return data.size();
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex) {
        if (rowIndex < data.size()) {
            switch (columnIndex) {
                case 0:
                    return data.get(rowIndex).getSelected();
                case 1:
                    return data.get(rowIndex).getKey();
                case 2:
                    return data.get(rowIndex).getValue();
                case 3:
                    return data.get(rowIndex).getDataType();
            }
        }
        return null;
    }

    //重写isCellEditable方法

    @Override
    public boolean isCellEditable(int rowIndex, int columnIndex) {
        return true;
    }

    //重写setValueAt方法
    @Override
    public void setValueAt(Object aValue, int rowIndex, int columnIndex) {
        if (rowIndex < data.size()) {
            switch (columnIndex) {
                case 0:
                    data.get(rowIndex).setSelected((Boolean) aValue);
                    break;
                case 1:
                    data.get(rowIndex).setKey((String) aValue);
                    break;
                case 2:
                    data.get(rowIndex).setValue((String) aValue);
                    break;
                case 3:
                    data.get(rowIndex).setDataType((BoundDataType) aValue);
                    break;
            }
        }
        this.fireTableCellUpdated(rowIndex, columnIndex);
    }


}

File: src/main/kotlin/com/developerphil/adbidea/ui/BoundTableModel.kt
package com.developerphil.adbidea.ui

import com.developerphil.adbidea.bean.BoundDataType
import com.developerphil.adbidea.bean.BoundItemBean
import java.util.*
import javax.swing.table.AbstractTableModel

class BoundTableModel : AbstractTableModel() {
    private val cellType = arrayOf<Class<*>>(Boolean::class.java, String::class.java, String::class.java, BoundDataType::class.java)
    private val title = arrayOf("select", "key", "value", "type")

    private val data = ArrayList<BoundItemBean>()


    fun getData(): List<BoundItemBean> {
        return data.filter { it.selected&& it.key.isNotEmpty()}
    }

    fun setData(data: List<BoundItemBean>) {
        this.data.clear()
        this.data.addAll(data)
        fireTableDataChanged()
    }

    fun clear() {
        val size = this.data.size
        this.data.clear()
        fireTableRowsDeleted(0, size)
    }

    fun addEmptyRow() {
        data.add(BoundItemBean(true, "", "", BoundDataType.STRING))
        fireTableRowsInserted(data.size - 1, data.size - 1)
    }

    fun removeRow(row: Int) {
        if (row > -1 && row < data.size) {
            data.removeAt(row)
        }
        fireTableRowsDeleted(row, row)
    }

    override fun getColumnClass(arg0: Int): Class<*> {
        return cellType[arg0]
    }

    override fun getColumnName(arg0: Int): String {
        return title[arg0]
    }

    override fun getColumnCount(): Int {
        return title.size
    }

    override fun getRowCount(): Int {
        return data.size
    }

    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any? {
        if (rowIndex < data.size) {
            when (columnIndex) {
                0 -> return data[rowIndex].selected
                1 -> return data[rowIndex].key
                2 -> return data[rowIndex].value
                3 -> return data[rowIndex].dataType
            }
        }
        return null
    }

    //重写isCellEditable方法

    override fun isCellEditable(rowIndex: Int, columnIndex: Int): Boolean {
        return true
    }

    //重写setValueAt方法
    override fun setValueAt(aValue: Any?, rowIndex: Int, columnIndex: Int) {
        if (rowIndex < data.size) {
            when (columnIndex) {
                0 -> data[rowIndex].selected = aValue as Boolean
                1 -> data[rowIndex].key = aValue as String
                2 -> data[rowIndex].value = aValue as String
                3 -> data[rowIndex].dataType = aValue as BoundDataType
            }
        }
        this.fireTableCellUpdated(rowIndex, columnIndex)
    }


}
